### PR TITLE
add ValueError in Euler for negative numbers

### LINKED ISSUE #8786: Euler number of negative integers
Currently `euler(-1)`, `euler(-3)`, etc. all evaluate to 0, as the first test in `euler.eval` is

``` python
        if m.is_odd:
            return S.Zero
```

but euler numbers are not defined on negative numbers, so this is wrong (and should probably raise a `ValueError`).

### PR DESCRIPTION
**Pull request by mvnnn**
fixes #8786

### CODE REVIEW

**Comment by mvnnn** [comment]
ping @asmeurer

**Comment by mvnnn** [comment]
ping @jksuom

File: sympy/functions/combinatorial/numbers.py [hunk 1]
```diff
@@ -733,13 +734,17 @@ class euler(Function):
...
     @classmethod
     def eval(cls, m):
-        if m.is_odd:
-            return S.Zero
-        if m.is_Integer and m.is_nonnegative:
-            from mpmath import mp
-            m = m._to_mpmath(mp.prec)
-            res = mp.eulernum(m, exact=True)
-            return Integer(res)
+        if (not m.is_Number) or m.is_nonnegative:
```
Comment (jksuom) [inline sympy/functions/combinatorial/numbers.py]: Why ``not m.is_Number``?

Comment (mvnnn) [inline sympy/functions/combinatorial/numbers.py]: it's allow to pass all ```Symbol```. which one is not ```negative```.

Comment (jksuom) [inline sympy/functions/combinatorial/numbers.py]: It will also allow other objects, e.g, Sets, that should also raise ValueError.

Comment (mvnnn) [inline sympy/functions/combinatorial/numbers.py]: @jksuom now, it's looks good.

Comment (mvnnn) [inline sympy/functions/combinatorial/numbers.py]: @jksuom Now, it's looks good.

**Comment by mvnnn** [comment]
@jksuom Now, it's looks good.

**Comment by jksuom** [comment]
I don't understand the role of the condition ``fuzzy_or(m.is_positive, m.is_negative)``. If its purpose is only to let symbols pass, perhaps something like ``m.is_integer`` would be more appropriate.

I also think that infinite values and ``nan`` should be rejected.

And please, don't squash the commits after each change. It makes it harder to review the PR.

**Comment by mvnnn** [comment]
```
fuzzy_or(m.is_positive, m.is_negative)
```
if ```m``` is ```positive``` and ```negative```both then it's return ```None```.

**Comment by mvnnn** [comment]
```
fuzzy_or(m.is_positive, m.is_negative) is None
```
@jksuom if ```m``` is not ```positive``` or not```negative``` then it's return ```None```.in other word if we declare ```Symbol``` or ```sets``` are positive or negative then it's not return ```None``` because above condition became false.

**Comment by mvnnn** [comment]
```
I also think that infinite values and nan should be rejected.
```
that's working in ```infinite``` and ```nan``` case.if we input ```infinite``` then it's return ```None``` and if we add ```infinite``` -->```negative``` then it's raise error and ```infinite```-->```positive``` then it's return it's corresponding value.i think i need to add test cases for ```infinite```. should i do ?

**Comment by mvnnn** [comment]
```
I also think that infinite values and nan should be rejected.
```
@jksuom that's working in ```infinite``` and ```nan``` case.if we input ```infinite``` then it's return ```None``` and if we add ```infinite``` -->```negative``` then it's raise error and ```infinite```-->```positive``` then it's return it's corresponding value.

**Comment by mvnnn** [comment]
@jksuom if you are still unclear check the test cases.

**Comment by jksuom** [comment]
I was expecting that you could explain the reasoning for this test:
``fuzzy_or([m.is_positive, m.is_negative]) is None``

It seems that it is satisfied for example if ``m.is_positive is False`` and ``m.is_negative is None``.
In this case I would expect that ValueError be returned except if m could be 0.

**Comment by jksuom** [comment]
We are interested in whether the argument is a non-negative integer or not. None is returned if this is possible but not sure. The fuzzy_or test does not help to decide this. For example, we get
```
>>> n = symbols('n', integer=True)
>>> m = n + S(1)/2
>>> euler(m)
euler(n + 1/2)
```
because of
```
>>> fuzzy_or([m.is_positive, m.is_negative]) is None
True
```
We should get ValueError, since we know that
```
>>> m.is_integer
False
```

**Comment by jksuom** [comment]
We want to find out if m is or can be a non-negative integer. So ``m.is_integer`` is obviously useful and even necessary. But I don't think that the code should be so complicated and hard to follow. For example, ``m.is_negative`` appears twice. Once should be enough, since the test may be time consuming. Moreover, there seems to be no obvious reason for calling ``m.is_rational``.

Can you try to describe your algorithm in plain text. Then it could be easier to see how the tests should be arranged.

**Comment by mvnnn** [comment]
```
I don't think that the code should be so complicated and hard to follow.
```
thanks @jksuom, i also think that. i will improve that.

File: sympy/functions/combinatorial/numbers.py [hunk 2]
```diff
@@ -733,13 +734,20 @@ class euler(Function):
...
+            elif m.is_odd and m >= 1:
+                return S.Zero
+            elif m.is_nonnegative:
+                from mpmath import mp
+                m = m._to_mpmath(mp.prec)
+                res = mp.eulernum(m, exact=True)
+                return Integer(res)
+        if m.is_negative:
+            raise ValueError('Euler numbers are not defined on negative numbers')
+        elif m.is_rational or m.is_positive is False:
```
Comment (mvnnn) [inline sympy/functions/combinatorial/numbers.py]: @jksuom Now, that's looks good. ```is_rational``` is valid because i cover ```positive```, ```negative``` and zero in above condition and we need to return ```ValueError``` except those numbers.

**Comment by jksuom** [comment]
There are basically two predicates that determine the result: ``m.is_integer`` and ``m.is_nonnegative``. Each one has three possible values: True, False and None. Altogether there are nine possible combinations. You should start by deciding what to do in each of these nine cases.

**Comment by mvnnn** [comment]
sorry for delay, currently i am busy to write GSOC proposal . we will solve this issue after 25th march.

**Comment by czgdp1807** [comment]
Closing this in reference too, https://github.com/sympy/sympy/issues/8786#issuecomment-565810037
